sklearn/preprocessing/tests/test_knn_imputation.py

```diff
@@ -0,0 +1,432 @@
+        [np.nan, 2],
+        [4,      3],
+        [5,      6],
+        [7,      7],
+        [9,      8],
+        [11,     10]
+    ])
+
+    # Test with "uniform" weight (or unweighted)
+    X_imputed_uniform = np.array([
+        [0,      0],
+        [5,      2],
+        [4,      3],
+        [5,      6],
+        [7,      7],
+        [9,      8],
+        [11,     10]
+    ])
+
+    imputer = KNNImputer(weights="uniform")
+    assert_array_equal(imputer.fit_transform(X), X_imputed_uniform)
+
+    # Test with "callable" weight
+    def no_weight(dist):
+        return None
+
+    imputer = KNNImputer(weights=no_weight)
+    assert_array_equal(imputer.fit_transform(X), X_imputed_uniform)
+
+    # Test with "distance" weight
+    nn = NearestNeighbors(metric="masked_euclidean")
+    nn.fit(X)
+    # Get distance of "n_neighbors" neighbors of row 1
+    dist, index = nn.kneighbors()
+    dist = dist[1, :]
+    index = index[1, :]
+    weights = 1 / dist
+    values = X[index, 0]
+    imputed = np.dot(values, weights) / np.sum(weights)
+
+    # Manual calculation
+    X_imputed_distance1 = np.array([
+        [0,                 0],
+        [3.850393700787402, 2],
+        [4,                 3],
+        [5,                 6],
+        [7,                 7],
+        [9,                 8],
+        [11,                10]
+    ])
+
+    # NearestNeighbor calculation
+    X_imputed_distance2 = np.array([
+        [0,                 0],
+        [imputed,           2],
+        [4,                 3],
+        [5,                 6],
+        [7,                 7],
+        [9,                 8],
+        [11,                10]
+    ])
+
+    imputer = KNNImputer(weights="distance")
+    assert_array_almost_equal(imputer.fit_transform(X), X_imputed_distance1,
+                              decimal=6)
+    assert_array_almost_equal(imputer.fit_transform(X), X_imputed_distance2,
+                              decimal=6)
+
+    # Test with weights = "distance" and n_neighbors=2
+    X = np.array([
+        [np.nan, 0,      0],
+        [2,      1,      2],
+        [3,      2,      3],
+        [4,      5,      5],
+    ])
+
+    statistics_mean = [3, 2, 2.5]
+
+    X_imputed = np.array([
+        [2.3828, 0,     0],
+        [2,      1,     2],
+        [3,      2,     3],
+        [4,      5,     5],
+    ])
+
+    imputer = KNNImputer(n_neighbors=2, weights="distance")
+    assert_array_almost_equal(imputer.fit_transform(X), X_imputed,
+                              decimal=4)
+    assert_array_equal(imputer.statistics_, statistics_mean)
+
+
+def test_metric_type():
+    X = np.array([
+        [0,      0],
+        [np.nan, 2],
+        [4,      3],
+        [5,      6],
+        [7,      7],
+        [9,      8],
+        [11,     10]
+    ])
+
+    # Test with a metric type without NaN support
+    imputer = KNNImputer(metric="euclidean")
+    assert_raises(ValueError, imputer.fit, X)
+
+
+def test_imputation_copy():
```

Comment:
Merge this with the equivalent Imputer test?

Reply:
I was thinking it might be more convenient, for debugging etc.,  if the tests for the two Imputers were kept separate in their respective "segments" of the file, esp. given that code re-use is trivial in any case. what do you think?

Reply:
I generally think that if we want multiple thing to have the same API, we should test that together.

Reply:
I generally agree with your point but these two transformers have so many other tests that cannot be easily merged, if at all, so only merging this one particular test might cause headaches down the line is what I am thinking. An alternative could be a "General tests" section within the file that is not specific to any particular Imputer. But if you still think it is worthwhile merging them, I will go ahead with that.